### PR TITLE
Simpler Survey deploy system.  (Debt reduction, prior to SHIELD)

### PR DESCRIPTION
**Pull request by gregglind**

### CODE REVIEW

File: package.json [hunk 1]
```diff
@@ -48,7 +48,8 @@
     "thirdparty": "python scripts/check-ui-tour-changes.py thirdparty/uitour.js",
     "serve": "serve -p 3111",
     "open-local": "opener 'http://localhost:3111/deploy/en-US/repair/?{\"runner\":{\"alwaysRun\":true}}'",
-    "open-prod": "opener 'https://self-repair.mozilla.org/en-US/repair/?{\"runner\":{\"alwaysRun\":true}}'"
+    "open-prod": "opener 'https://self-repair.mozilla.org/en-US/repair/?{\"runner\":{\"alwaysRun\":true}}'",
+    "surveys": "TRAVIS=1 NPM_CONFIG_KARMAFILES=\"test/**/*deployed-survey-config.js\" npm run test"
```
Comment (gregglind) [inline package.json]: this is new.

File: src/common/rules.js [hunk 2]
```diff
@@ -0,0 +1,42 @@
...
+  globalstrict:true, nomen:false, newcap:true, esnext: true, moz: true,
+  asi: true  */
+
+/*global module */
+
+function isRe (thing) {
+  return Object.prototype.toString.call( thing ) == "[object RegExp]"
+}
+
+function matchRulePart(subObj, expected) {
```
Comment (gregglind) [inline src/common/rules.js]: only regex and 'everything else'.  No functions.

NOT SUPPORTED:
- thing.level1.level2

File: src/recipes/heartbeat-by-user-first-impression/config.js [hunk 3]
```diff
@@ -12,15 +12,71 @@
 
 "use strict";
 
-let million = Math.pow(10,6);
-let thousand = Math.pow(10,3);
-let percent = 0.01;
-let days = 24 * 60 * 60 * 1000;
+const VERSION=33;
```
Comment (gregglind) [inline src/recipes/heartbeat-by-user-first-impression/config.js]: version come here.

File: src/recipes/heartbeat-by-user-first-impression/config.js [hunk 4]
```diff
@@ -32,25 +88,51 @@ let supportedLocales = [ 'de',
...
 
+
+const million = Math.pow(10,6);
+const thousand = Math.pow(10,3);
+const percent = 0.01;
+const days = 24 * 60 * 60 * 1000;
+
+
+// convenience function
+function asBreaks(ratios) {
```
Comment (gregglind) [inline src/recipes/heartbeat-by-user-first-impression/config.js]: this gets hoisted in js.

File: src/recipes/heartbeat-by-user-first-impression/index.js [hunk 5]
```diff
@@ -13,6 +13,8 @@
 "use strict";
 let common = require("../../common");
 let allconfigs = require("./config");
+let { getEngagementUrl } = require("./utils");
```
Comment (gregglind) [inline src/recipes/heartbeat-by-user-first-impression/index.js]: introduces utils

File: src/recipes/heartbeat-by-user-first-impression/index.js [hunk 6]
```diff
@@ -48,7 +51,7 @@ let { hasAny } = require("../../jetpack/array");
 */
 
 const NAME="heartbeat by user v1";
-const VERSION=32;
+const VERSION=allconfigs.VERSION;
```
Comment (gregglind) [inline src/recipes/heartbeat-by-user-first-impression/index.js]: pulls in from config

File: src/recipes/heartbeat-by-user-first-impression/index.js [hunk 7]
```diff
@@ -114,7 +116,7 @@ let shouldRun = function (userstate, config, extras) {
   let lastRun = extras.lastRun || data.lastRun || 0;
   let locale = (userstate.locale || extras.locale || "unknown").toLowerCase();
 
-  config = config || allconfigs[channel];
+  config = config || allconfigs.sampling[channel];
```
Comment (gregglind) [inline src/recipes/heartbeat-by-user-first-impression/index.js]: deeper nesting

File: src/recipes/heartbeat-by-user-first-impression/index.js [hunk 8]
```diff
@@ -224,49 +228,10 @@ let run = function (state, extras) {
     }
   };
 
-  //      updateChannel:  "unkown",
-  //    fxVersion: "unknown",
-
-  // as shown below, this is en-us only!
   //let engagementUrl =  `https://www.mozilla.org/en-US/firefox/feedback/?updateChannel=${state.updateChannel}&fxVersion=${state.fxVersion}`;  //"http://localhost/enagement.html",
+  let eOpts = extend(state, {VERSION: VERSION, locale: locale})
```
Comment (gregglind) [inline src/recipes/heartbeat-by-user-first-impression/index.js]: modifies the format here a little, to put *everything* only one level deep.

File: src/recipes/heartbeat-by-user-first-impression/index.js [hunk 9]
```diff
@@ -224,49 +228,10 @@ let run = function (state, extras) {
...
+  let engagementUrl = getEngagementUrl(eOpts, allconfigs.engagementRules);
 
-  let eUrls = [
-    //`https://qsurvey.mozilla.com/s3/Firefox-User-Sentiment/?updateChannel=${state.updateChannel}&fxVersion=${state.fxVersion}`
-  ];
-
-  let cutBreaks = function (arr, breaks, rng=Math.random()) {
```
Comment (gregglind) [inline src/recipes/heartbeat-by-user-first-impression/index.js]: cutBreaks and getEngagementUrl  go to utils.

File: src/recipes/heartbeat-by-user-first-impression/utils.js [hunk 10]
```diff
@@ -0,0 +1,50 @@
+/* This Source Code Form is subject to the terms of the Mozilla Public
```
Comment (gregglind) [inline src/recipes/heartbeat-by-user-first-impression/utils.js]: new file.

File: test/common/test-rules.js [hunk 11]
```diff
@@ -0,0 +1,114 @@
+/* This Source Code Form is subject to the terms of the Mozilla Public
```
Comment (gregglind) [inline test/common/test-rules.js]: new file, tests for rules.

File: test/recipes/test-heartbeat-by-user-first-impression.js [hunk 12]
```diff
@@ -32,6 +32,7 @@ let personinfo = require("../../src/common/personinfo");
 
 let R = require("../../src/recipes/heartbeat-by-user-first-impression");
 let C = require("../../src/recipes/heartbeat-by-user-first-impression/config");
+let U = require("../../src/recipes/heartbeat-by-user-first-impression/utils");
```
Comment (gregglind) [inline test/recipes/test-heartbeat-by-user-first-impression.js]: gross, going to do all the tests for the utils and config in here too.

File: test/recipes/test-heartbeat-by-user-first-impression.js [hunk 13]
```diff
@@ -117,7 +147,7 @@ describe("heartbeat-by-user-first-impression", function () {
         var allchannels = ['nightly', 'aurora', 'beta', 'release'];
 
         it('config has right channels: '+ allchannels.join('|'), () => {
-          expect(C).keys(allchannels);
+          expect(C.sampling).keys(allchannels);
```
Comment (gregglind) [inline test/recipes/test-heartbeat-by-user-first-impression.js]: match format of the new config file.

File: test/recipes/test-heartbeat-by-user-first-impression.js [hunk 14]
```diff
@@ -466,3 +496,107 @@ describe("heartbeat-by-user-first-impression", function () {
     });
   });
 })
+
+describe("hb utils", function () {
```
Comment (gregglind) [inline test/recipes/test-heartbeat-by-user-first-impression.js]: new tests for the utils file.

File: test/test-built/test-deployed-survey-config.js [hunk 15]
```diff
@@ -0,0 +1,130 @@
+/* This Source Code Form is subject to the terms of the Mozilla Public
```
Comment (gregglind) [inline test/test-built/test-deployed-survey-config.js]: this test just does reporting on survey config.  No actual test.

File: src/recipes/heartbeat-by-user-first-impression/config.js [hunk 16]
```diff
@@ -12,15 +12,71 @@
...
+
+const engagementRules = [
+  // en, aurora.
+  {
+    alias: '^en, ^aurora',
+    rule: {
+      locale: /^en/,
+      channel: /^aurora/
+    },
+    urls: [
```
Comment (rayborn) [inline src/recipes/heartbeat-by-user-first-impression/config.js]: I mentioned before, that I think that this should be objects to make sure url and break is 1:1 (not sure if you missed it or just disagreed)
urls [{url: <>, break: <>}, {url: <>, break: <>}, ... ]

Perhaps breaks should be sample pcts?  Specifying 20, 30, and 50 might be more intuitive than 20, 50, and 100.  Although we would need to test that sum(breaks) <= 100

As it stands do we test that break(n) < break(n-1)?  <ight be a decent TODO.

File: src/recipes/heartbeat-by-user-first-impression/config.js [hunk 17]
```diff
@@ -12,15 +12,71 @@
...
 
-/*
-  days since last ask?  -- function or constant?
+/** EngagementUrl Rules
 
-*/
+  Notes:
+  1.  first match will win.  PUT MORE SPECIFIC RULES FIRST.
+      Note: NO automated check of rule order.
+  2.  'qsurvey' urls will postpend all the right query args.  others won't.
```
Comment (rayborn) [inline src/recipes/heartbeat-by-user-first-impression/config.js]: Nice!

File: src/recipes/heartbeat-by-user-first-impression/config.js (hunk 4 above)
Comment (rayborn) [inline src/recipes/heartbeat-by-user-first-impression/config.js]: This is nice, a little documentation would be good.

File: test/recipes/test-heartbeat-by-user-first-impression.js [hunk 18]
```diff
@@ -44,6 +45,38 @@ let sendTourEvent = function (aEventName, aData) {
...
 
+/**
+  * same length, increasing, 1 for right endpoint
+  */
+function okEngagementSpec(spec, name) {
+  let {urls, breaks} = spec;
+  let msg = `${name} wrong length of breaks`;
+  expect(urls.length, msg).equal(breaks.length);
+  msg = `are not strictly increasing: ${name}, ${breaks}`;
+  expect(U._increasing(breaks), msg).to.be.true();
```
Comment (rayborn) [inline test/recipes/test-heartbeat-by-user-first-impression.js]: Nice

File: src/recipes/heartbeat-by-user-first-impression/config.js (hunk 16 above)
Comment (gregglind) [inline src/recipes/heartbeat-by-user-first-impression/config.js]: This is supported.  I still don't like the unified in objects, but at line 42, 'asBreaks' is shown, which does that rescale :)

File: src/recipes/heartbeat-by-user-first-impression/config.js [hunk 19]
```diff
@@ -12,15 +12,71 @@
...
+  {
+    alias: '^en, ^aurora',
+    rule: {
+      locale: /^en/,
+      channel: /^aurora/
+    },
+    urls: [
+      "https://qsurvey.mozilla.com/s3/Developer-Audience-Survey-V2/"
+    ],
+    breaks: asBreaks([
```
Comment (gregglind) [inline src/recipes/heartbeat-by-user-first-impression/config.js]: @rayborn , right here... asBreaks(list_of_ratios) does what you ask.

File: src/recipes/heartbeat-by-user-first-impression/config.js (hunk 4 above)
Comment (gregglind) [inline src/recipes/heartbeat-by-user-first-impression/config.js]: will do.

File: src/common/rules.js [hunk 20]
```diff
@@ -0,0 +1,42 @@
...
+/*jshint forin:true, noarg:false, noempty:true, eqeqeq:true, bitwise:true,
+  strict:true, undef:true, curly:false, browser:true,
+  unused:true, indent:2, maxerr:50, devel:true, node:true, boss:true, white:true,
+  globalstrict:true, nomen:false, newcap:true, esnext: true, moz: true,
+  asi: true  */
+
+/*global module */
+
+function isRe (thing) {
+  return Object.prototype.toString.call( thing ) == "[object RegExp]"
```
Comment (mythmon) [inline src/common/rules.js]: Would this be better as `thing instanceof RegExp`?

**Pull request merged**